src/util/mpsc.rs

```diff
@@ -0,0 +1,267 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*!
+
+This module provides an implementation of mpsc channel based on
+crossbeam_channel. Comparing to the crossbeam_channel, this implementation
+supports closed detection and try operations.
+
+*/
+
+use crossbeam_channel;
+use std::sync::atomic::{AtomicIsize, Ordering};
+use std::sync::mpsc::{RecvError, RecvTimeoutError, SendError, TryRecvError, TrySendError};
+use std::sync::Arc;
+use std::time::Duration;
+
+struct RefCount {
+    sender: AtomicIsize,
+    receiver: AtomicIsize,
+}
+
+impl RefCount {
+    fn new() -> RefCount {
+        RefCount {
+            sender: AtomicIsize::new(1),
+            receiver: AtomicIsize::new(1),
+        }
+    }
+
+    #[inline]
+    fn is_receiver_closed(&self) -> bool {
+        self.receiver.load(Ordering::Acquire) == 0
+    }
+
+    #[inline]
+    fn is_sender_closed(&self) -> bool {
+        self.sender.load(Ordering::Acquire) == 0
+    }
+}
+
+pub struct Sender<T> {
+    sender: crossbeam_channel::Sender<T>,
+    replica_cnt: Arc<RefCount>,
+}
+
+impl<T> Clone for Sender<T> {
+    #[inline]
+    fn clone(&self) -> Sender<T> {
+        self.replica_cnt.sender.fetch_add(1, Ordering::AcqRel);
+        Sender {
+            sender: self.sender.clone(),
+            replica_cnt: self.replica_cnt.clone(),
+        }
+    }
+}
+
+impl<T> Drop for Sender<T> {
```

Comment:
What happens to `T`s when `Sender` and `Receive` are dropped?

Reply:
They will be dropped.